src/stan/services/pathfinder/single.hpp

```diff
@@ -0,0 +1,1104 @@
+          << " ";
+      msg << " " << std::setw(10) << std::setprecision(4) << lbfgs.alpha0()
+          << " ";
+      msg << " " << std::setw(7) << lbfgs.grad_evals() << " ";
+      msg << " " << lbfgs.note() << " ";
+      logger.info(msg.str());
+    }
+
+    if (lbfgs_ss.str().length() > 0) {
+      logger.info(lbfgs_ss);
+      lbfgs_ss.str("");
+    }
+    /*
+     * If the retcode is -1 then linesearch failed even with a hessian reset
+     * so the current vals and grads are the same as the previous iter
+     * and we are exiting
+     */
+    if (likely(ret != -1)) {
+      param_buff.push_back(lbfgs.curr_x() - prev_params);
+      grad_buff.push_back(lbfgs.curr_g() - prev_grads);
+      prev_params = lbfgs.curr_x();
+      prev_grads = lbfgs.curr_g();
+      current_history_size = std::min(current_history_size + 1,
+                                      static_cast<std::size_t>(history_size));
+      if (internal::check_curve(param_buff.back(), grad_buff.back(), logger)) {
+        alpha = internal::form_diag(alpha, grad_buff.back(), param_buff.back());
+      }
+      Eigen::Map<Eigen::MatrixXd> Ykt_map(Ykt_mat.data(), num_parameters,
+                                          current_history_size);
+      for (Eigen::Index i = 0; i < current_history_size; ++i) {
+        Ykt_map.col(i) = grad_buff[i];
+      }
+      Eigen::Map<Eigen::MatrixXd> Skt_map(Skt_mat.data(), num_parameters,
+                                          current_history_size);
+      for (Eigen::Index i = 0; i < current_history_size; ++i) {
+        Skt_map.col(i) = param_buff[i];
+      }
+      std::string iter_msg(path_num + "Iter: ["
+                           + std::to_string(lbfgs.iter_num()) + "] ");
+      if (STAN_DEBUG_PATH_ITERS) {
+        logger.info(iter_msg + "\n------------ Iter: "
+                    + std::to_string(lbfgs.iter_num()) + "------------\n");
+      }
+
+      auto pathfinder_res = internal::pathfinder_impl(
+          rng, lp_fun, constrain_fun, alpha, lbfgs.curr_x(), lbfgs.curr_g(),
+          Ykt_map, Skt_map, num_elbo_draws, iter_msg, logger);
+      num_evals += pathfinder_res.first.fn_calls;
+      if (pathfinder_res.first.elbo > elbo_best.elbo) {
+        elbo_best = std::move(pathfinder_res.first);
+        taylor_approx_best = std::move(pathfinder_res.second);
+        best_E = lbfgs.iter_num();
+      }
+      if (refresh > 0
+          && (lbfgs.iter_num() == 0 || (lbfgs.iter_num() % refresh == 0))) {
+        logger.info(iter_msg + ": ELBO ("
+                    + std::to_string(pathfinder_res.first.elbo) + ")");
+      }
+    }
+    if (msg.str().length() > 0) {
+      logger.info(msg);
+    }
+    if (save_iterations) {
+      diagnostic_writer(std::make_tuple(lbfgs.curr_x(), lbfgs.curr_g()));
+    }
+  }
+  int return_code = error_codes::OK;
+  if (ret >= 0) {
+    logger.info("Optimization terminated normally: ");
+  } else {
+    logger.info("Optimization terminated with error: ");
```

Comment:
A single error message should all go on one line, so just a + here with the message.

Reply:
👍